File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */



// APPROACH: SLIDING WINDOW APPROACH
// APPROACH: window starts from the left and expands to the right, keeping track of the minimum price encountered and updating the maximum profit 

// the number zero is being assigned to the variable 'left', representing the index of the leftmost element in the sliding window
// the number one is being assigned to the variable 'right', representing the element to the right of the window
// the number zero is being assigned to the variable 'max', indicating profit and no profit has been made yet

// code enters a while loop that continues as long as the 'right' index is less than the length of the 'prices' array

// inside the loop, the code checks if the price at 'prices[right]' is less than or equal to the price at 'prices[left]'
// if it is, it means we can slide the window to the right and update the 'left' index to 'right' (makes sures that 'left' always points to the lowest price encountered so far)

// next, calculate the window's size by subtracting prices[left] from prices[right]
// use Math.max to update the 'max' variable, ensuring 'max' always holds the maximum profit 
// finally, right index is incremented to move the sliding window to the right

// once loop finishesm return max



// var maxProfit = function(prices) {
//     let left = 0;
//     let right = 1;
//     let max = 0;
    
//     while (right < prices.length) {
//         if (prices[right] <= prices[left]) {
//             left = right;
//         }
        
//         const window = prices[right] - prices[left];
        
//         max = Math.max(max, window);
//         right++;
//     }
//     return max; 
// }




// APPROACH: BUY LOW, SELL HIGH
// APPROACH: by keeping track of the minium price and calculating the potential profit at each step, it finds the maximum possible profit from buying and selling the stock

var maxProfit = function(prices) {
    let minPrice = Infinity; // set to Infinity to ensure that it will be updated with the lowest price encountered
    let maxProfit = 0; // set to zero, meaning no profit has been made yet
    
    for (let i = 0; i < prices.length; i++) { // loop through the prices array, starting from first element (i = 0) and continues until the end (i < prices.length)
        if (prices[i] < minPrice) { // check to see if the current price is less than the current 'minPrice'
            minPrice = prices[i]; // if less than 'minPrice', 'minPrice' is updated to the current price. ensures 'minPrice' always holds the lowest price encountered so far
        } else if (prices[i] - minPrice > maxProfit) { // if current price is not less than 'minPrice', there is potential selling opportunity. Calc profit by subtracting 'minPrice' from current price
            maxProfit = prices[i] - minPrice; // if profit is greater than current 'maxProfit', update 'maxProfit'
        }
    }
    return maxProfit; // return maxProfit
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int minPrice = Integer.MAX_VALUE;
        // initialize the minimum price to a very large number

        int maxProfit = 0;
        // initialize the maximum profit to zero

        // loop through each price using an index
        for (int i = 0; i < prices.length; i++) {
            int currentPrice = prices[i]; // current day's price

            // if we find a cheaper price, update minPrice
            if (currentPrice < minPrice) {
                minPrice = currentPrice;
            }
            // otherwise, calculate the profit if we sold today
            else {
                int profit = currentPrice - minPrice; 

                // If this profit is greater than the previous max, update it
                if (profit > maxProfit) {
                    maxProfit = profit;
                }
            }
        }

        // return the best profit found
        return maxProfit; 
    }
}

// walk through days left to right and keep two notes
// 1. cheapest price we've seen so far (best day to have bought up to now)
// 2. best profit we could have made by selling today (price today - cheapest so far), and update the global best if it's higher

// BIG-O
// Time Complexity: O(n)
// One pass, constant work per element. No nested loops or sorting
// Space Complexity: O(1)
// Store just two integers - minimum so far and best profit

// Any approach that compares each pair is O(n^2). Sorting is O(n log n) but breaks the buy-before-sell constraint. One-pass min-so-far preserves order and minimizes time.

// "Scan once while tracking the cheapest price seen so far. For each day, compute the profit if I sold today and update a running maximum."
// "That's O(n) time, O(1) space. Sorting doesn't work because it loses orer, and brute force is O(n^2)"
// "This approach preserves the buy-before-sell constraint"